gtsam/slam/PlanarProjectionFactor.h

```diff
@@ -0,0 +1,318 @@
+/**
+ * ProjectionFactor, but with pose2 (robot on the floor)
+ *
+ * This factor is useful for high-school robotics competitions,
+ * which run robots on the floor, with use fixed maps and fiducial
+ * markers.
+ *
+ * @see https://www.firstinspires.org/
+ *
+ * @file PlanarProjectionFactor.h
+ * @brief for planar smoothing
+ * @date Dec 2, 2024
+ * @author joel@truher.org
+ */
+#pragma once
+
+#include <gtsam/base/Testable.h>
+#include <gtsam/base/Lie.h>
+
+#include <gtsam/geometry/Cal3DS2.h>
+#include <gtsam/geometry/PinholeCamera.h>
+#include <gtsam/geometry/Point3.h>
+#include <gtsam/geometry/Pose2.h>
+#include <gtsam/geometry/Pose3.h>
+#include <gtsam/geometry/Rot3.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+#include <gtsam/base/numericalDerivative.h>
+
+
+namespace gtsam {
+
+    /**
+     * @class PlanarProjectionFactorBase
+     * @brief Camera projection for robot on the floor.  Measurement is camera pixels.
+     */
+    class PlanarProjectionFactorBase {
+    protected:
+        PlanarProjectionFactorBase() {}
+
+        /**
+         * @param measured corresponding point in the camera frame
+         * @param poseKey index of the robot pose in the z=0 plane
+         */
+        PlanarProjectionFactorBase(const Point2& measured) : measured_(measured) {}
+
+        /**
+         * @param landmark point3
+         * @param pose
+         * @param offset oriented parallel to pose2 zero i.e. +x
+         * @param calib
+         * @param H1 landmark jacobian
+         * @param H2 pose jacobian
+         * @param H3 offset jacobian
+         * @param H4 calib jacobian
+         */
+        Point2 h(
+            const Point3& landmark,
+            const Pose2& pose,
+            const Pose3& offset,
+            const Cal3DS2& calib,
+            OptionalMatrixType H1, // 2x3 (x, y, z)
+            OptionalMatrixType H2, // 2x3 (x, y, theta)
+            OptionalMatrixType H3, // 2x6 (rx, ry, rz, x, y, theta)
+            OptionalMatrixType H4  // 2x9
+        ) const {
+            try {
+                // this is x-forward z-up
+                gtsam::Matrix H0; // 6x6
+                Pose3 offset_pose = Pose3(pose).compose(offset, H0);
+                // this is z-forward y-down
+                gtsam::Matrix H00; // 6x6
+                Pose3 camera_pose = offset_pose.compose(CAM_COORD, H00);
+                PinholeCamera<Cal3DS2> camera = PinholeCamera<Cal3DS2>(camera_pose, calib);
+                if (H2 || H3) {
+                    // Dpose is for pose3, 2x6 (R,t)
+                    gtsam::Matrix Dpose;
+                    Point2 result = camera.project(landmark, Dpose, H1, H4);
```

Comment:
I'd probably do
```c++                    
const Point2 predicted = camera.project(landmark, Hpose ? H_wTc : nullptr, Hlandmark, Hcalib);
```

Reply:
You mean instead of if (HwTb || HbTc) ?

Reply:
You're right, that might already be handled. The key is to examine your logic to make sure no Jacobians are computed unless needed.